### ISSUE TITLE
How to find out the owners of Regions?

### ISSUE THREAD

**Issue by GorNishanov** [GabrielDosReis/ipr#268]
I am putting the owner back. However more thinking is required.
Owner is not set for Callable_species because it is not an Expr. 
Initial motivation for removal needs to be checked against use cases.

### Discussed in https://github.com/GabrielDosReis/ipr/discussions/254

<div type='discussions-op-text'>

<sup>Originally posted by **Xazax-hun** May 11, 2022</sup>
In https://github.com/GabrielDosReis/ipr/commit/59afd4298367bcd8a5ed0fd4efe1257ceb4a5f4d, `Region::owner()` was removed. We used this to check the context of the declarations. E.g., to check if a `Var` is a global, we traversed the parents  of the enclosing `Region`s and checked what kind of regions do we encounter. E.g., if we encounter a the `Region` of a `Block`, we know this is not a global variable. With  `Region::owner()` removed, we can no longer check what language construct do the parent regions of a `Var` correspond to. 

Is there an alternative way to get this information that I'm missing? Or are we supposed to maintain a mapping independently of IPR on the side?</div>

**Comment by GabrielDosReis** [comment]
`Region::owner()` being an `Expr` or an `Optional<Expr>` is indeed unsatisfactory.  Its semantics sounds off for nested regions, e.g.
```
int f(int i) {
    int x = i + 1;
   {                                      // this region does not have an owner
       int i = x - 1;
      x = i * 3;
   }
  return x;
}
```

**Comment by GorNishanov** [comment]
Block is a proper owner of the region, no?

```c++
int f(int i) { // block 1
    int x = i + 1;
   { // block 2. this region DOES have an owner. The owner is block 2
       int i = x - 1;
      x = i * 3;
   }
  return x;
}
```

**Comment by GabrielDosReis** [comment]
> Block is a proper owner of the region, no?

Yes, you're right.  I gave the wrong example.  The `Callable_species` is a better example.

Yes, before a _declarator_ (or _abstract declarator_) is elaborated, there is no `Expr` to attach the parameter list to.

**Comment by GabrielDosReis** [comment]
Resolved by #269
